src/server/main_service.cc

```diff
@@ -879,35 +918,41 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
 void Service::Discard(CmdArgList args, ConnectionContext* cntx) {
   RedisReplyBuilder* rb = (*cntx).operator->();
 
-  if (cntx->conn_state.exec_state == ConnectionState::EXEC_INACTIVE) {
+  if (!cntx->conn_state.exec_info.IsActive()) {
     return rb->SendError("DISCARD without MULTI");
   }
 
-  cntx->conn_state.exec_state = ConnectionState::EXEC_INACTIVE;
-  cntx->conn_state.exec_body.clear();
-
+  cntx->conn_state.exec_info.Clear();
   rb->SendOk();
 }
 
 void Service::Exec(CmdArgList args, ConnectionContext* cntx) {
   RedisReplyBuilder* rb = (*cntx).operator->();
 
-  if (cntx->conn_state.exec_state == ConnectionState::EXEC_INACTIVE) {
+  if (!cntx->conn_state.exec_info.IsActive()) {
     return rb->SendError("EXEC without MULTI");
   }
 
-  if (cntx->conn_state.exec_state == ConnectionState::EXEC_ERROR) {
-    cntx->conn_state.exec_state = ConnectionState::EXEC_INACTIVE;
-    cntx->conn_state.exec_body.clear();
+  if (cntx->conn_state.exec_info.state == ConnectionState::ExecInfo::EXEC_ERROR) {
+    cntx->conn_state.exec_info.Clear();
     return rb->SendError("-EXECABORT Transaction discarded because of previous errors");
   }
 
-  VLOG(1) << "StartExec " << cntx->conn_state.exec_body.size();
-  rb->StartArray(cntx->conn_state.exec_body.size());
-  if (!cntx->conn_state.exec_body.empty()) {
+  auto& exec_info = cntx->conn_state.exec_info;
+
+  // TODO: check watched keys for expiry
+  if (exec_info.watched_dirty.load()) {
+    exec_info.Clear();
+    return rb->SendNull();
+  }
+
+  VLOG(1) << "StartExec " << exec_info.body.size();
+  rb->StartArray(exec_info.body.size());
+
+  if (!exec_info.body.empty()) {
```

Comment:
The bigger problem is that the checks should be effectively part of the transaction, in a sense that there should be no changes between the checks and the actual execution.

Reply:
Wow, real kudos on noticing this!!!
I noticed this too but I am still contemplating whether it breaks strict serializability guarantees.

Reply:
I does for sure.
Nothing prevents a modify operation to be scheduled in-between the check and real exec. If it changes a watched key, then the execution is incorrect.

Reply:
You are absolutely correct. It was late at night and I was thinking about a corner case where watched keys and transaction keys do not overlap. In that case, it's more ambiguous. In any case, the only way to guarantee strict serializability is when transaction enters the execution phase. Which is well-defined by a state machine in shard-threads and not in coordination thread. The simplest way I can think of doing this is to introduce another hop before doing the transaction steps that serves as an barrier.